front/lib/api/assistant/agent.ts

```diff
@@ -398,66 +371,204 @@ export async function getNextAction(
   );
 
   if (res.isErr()) {
-    return new Err(
-      new Error(
-        `Error running use-tools action: [${res.error.type}] ${res.error.message}`
-      )
-    );
+    yield {
+      type: "agent_error",
+      created: Date.now(),
+      configurationId: agentConfiguration.sId,
+      messageId: agentMessage.sId,
+      error: {
+        code: "use_tools_error",
+        message: `Error running multi-actions agent action: [${res.error.type}] ${res.error.message}`,
+      },
+    } satisfies AgentErrorEvent;
+    return;
   }
 
   const { eventStream } = res.value;
 
   const output: {
-    name?: string;
-    arguments?: Record<string, string | boolean | number>;
-  } = {};
-
-  for await (const event of eventStream) {
-    if (event.type === "error") {
-      return new Err(
-        new Error(`Error generating action inputs: ${event.content.message}`)
-      );
-    }
+    name: string | null;
+    arguments: Record<string, string | boolean | number> | null;
+    generation: string | null;
+  } = {
+    name: null,
+    arguments: null,
+    generation: null,
+  };
+
+  let shouldYieldCancel = false;
+  let lastCheckCancellation = Date.now();
+  const redis = await redisClient();
+  let isGeneration = true;
+
+  try {
+    const _checkCancellation = async () => {
```

Comment:
What's this?

Reply:
same as in runGeneration, allows to react to cancellation from UI (it's cargo-culted from there)